add checkPermissions() to ensure kubeadm running in root

maxwell92 opened the pull request:
add checkPermissons() to ensure kubeadm running in root.  It's a TODO in cmd/kubeadm.go

<!-- Reviewable:start -->

---

This change is [<img src="https://reviewable.kubernetes.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/33902)

<!-- Reviewable:end -->

resouer committed:
Add no admit on node side

Update generated code

Refactored predicates & restore helper

justinsb committed:
Refactor: separate KubeletClient & ConnectionInfoGetter concepts

KubeletClient implements ConnectionInfoGetter, but it is not a complete
implementation: it does not set the kubelet port from the node record,
for example.

By renaming the method so that it does not implement the interface, we
are able to cleanly see where the "raw" GetConnectionInfo is used (it is
correct) and also have go type-checking enforce this for us.

Commit:
Switch kubectl to use watch.Until

jbeda committed:
Remove support for boot2docker

jbeda committed:
Make update-bindata be incremental build friendly

jbeda committed:
Use rsync to get source into build container

We also add "version" to all docker images and containers

This version is to be incremented manually when we change the shape of the build
image (like changing the golang version or the set of volumes in the data
container).  This will delete all older versions of images and containers when
the version is different.

jbeda committed:
Move release code (bash) out of build/common.sh

jbeda committed:
Make update-generated-*.sh not hack around with docker images

make-generated-{protobuf,runtime}.sh was doing some really nasty stuff with how
the build container was managed in order to copy results out.  Since we have
more flexibility to grab results out of the build container, we can now avoid
all of this.  Ideally we wouldn't have `hack` calling `build` at all, but we
aren't there yet.

wojtek-t committed:
Don't set timeouts in clients in tests

jeffvance committed:
Document selinux considerations

Commit:
Add support for vpshere cloud provider in kubeup

When deploying on vSphere using kube up add configuration
for vSphere cloud provider.

Commit:
Setup e2e test for no admit

juanvallejo committed:
handle invalid client config option errors

This patch provides a more relevant error message when a client
configuration option is passed with an invalid or non-existent value.

`$ kubectl get pods --cluster="non-existent"`
```
error: No configuration file found, please login or point to an existing
file
```

`$ kubectl get pods --cluster="non-existent"`
```
error: cluster "non-existent" does not exist
```

Commit:
Removing the Storage plugin from the list

smarterclayton committed:
DeltaFIFO should not report HasSynced until deletes are reported

Deletions identified by knownObjects on the initial sync replace
(immediately after the List from a reflector) should be considered part
of the initial sync since we have the information available at the time
we do the deletion.

An error during Replace() can result in Populated not being set, but it
was incorrect before (population would be wrong) and queueActionLocked
does not error except on "my cache is broken because I gave an incorrect
keyFunc".

dshulyak committed:
Delete evicted pet

If pet was evicted by kubelet - it will stuck in this state forever.
By analogy to regular pod we need to re-create pet so that it will
be re-scheduled to another node, so in order to re-create pet
and preserve consitent naming we will delete it in petset controller
and create after that.

Change-Id: Ib98bf7f34b3f2ab1582b9de34b5f4c5f84cd5215

jbeda committed:
Install test subpackages too

We weren't getting incremental builds because of new test only subpackages.  Our
voodoo combo of 'go install' and 'go test -c' didn't cache things like
'test/e2e_node/services'.  Add the '-i' flag to 'go test' to install test only
dependencies too.

dshulyak committed:
Verify evicted pods managed by petset controller will be recreated

Spawn pet set with single replica and simple pod. They will have
conflicting hostPort definitions, and spawned on the same node.

As the result pet set pod, it will be created after simple pod, will be
in Failed state. Pet set controller will try to re-create it. After
verifying that pet set pod failed and was recreated atleast once, we will
remove pod with conflicting hostPort and wait until pet set pod will be in
running state.

Change-Id: I5903f5881f8606c696bd390df58b06ece33be88a

Commit:
Update salt usage to accelerate kube-up.

Use state.show_highstate to wait for init highstate functions to be done on each node.
Then start highstate in parallel on all the minions.

Commit:
manual changes to let client-gen use versioned options

Commit:
Work around go cover bug

This avoids the whole command failing because of errors like the following:
```
# cover k8s.io/kubernetes/pkg/client/restclient
cover: internal error: block 268 overlaps block 270
```
For vendor code, igone
For local code replace the switch with an if statement

liggitt committed:
kubelet authentication/authorization proposal

sjenning committed:
kubelet: eviction: avoid duplicate action on stale stats

0xmichalis committed:
controller: scale proportionally before rolling out new templates

bprashanth committed:
Cleanup netexec image and usage.

mikedanese committed:
implement kubectl apply --prune

jszczepkowski committed:
HA master: add firewall rule for etcd cluster.

HA master: add firewall rule for accessing & securing etcd cluster.

jbeda committed:
Update comment on release.sh to reflect new release repo

dcbw committed:
kubenet: make it more apparent that kubenet ignores PodCIDR updates

https://github.com/kubernetes/kubernetes/issues/32900

vishh committed:
fix base image pinning during upgrades via cluster/gce/upgrade.sh

Signed-off-by: Vishnu kannan <vishnuk@google.com>

dashpole committed:
Split NodeDiskPressure into NodeInodePressure and NodeDiskPressure

guangxuli committed:
remove the tmp file

dashpole committed:
addressed comments

deads2k committed:
use service accounts as clients for controllers

juanvallejo committed:
Add option to set a service nodeport

This patch adds the option to set a nodeport when creating a NodePort
service. In case of a port allocation error due to a specified port
being out of the valid range, the error now includes the valid
range. If a `--node-port` value is not specified, it defaults to zero, in
which case the allocator will default to its current behavior of
assigning an available port.

This patch also adds a new helper function in `cmd/util/helpers.go` to
retrieve `Int32` cobra flags.

**Example**
```
$ kubectl create service nodeport mynodeport --tcp=8080:7777 --node-port=1
The Service "mynodeport" is invalid: spec.ports[0].nodePort: Invalid
value: 1: provided port is not in the valid range. Valid ports range
from 30000-32767

$ kubectl create service nodeport mynodeport --tcp=8080:7777 --node-port=30000
service "mynodeport" created

$ oc describe service mynodeport
Name:                   mynodeport
Namespace:              default
Labels:                 app=mynodeport
Selector:               app=mynodeport
Type:                   NodePort
IP:                     172.30.81.254
Port:                   8080-7777       8080/TCP
NodePort:               8080-7777       30000/TCP
Endpoints:              <none>
Session Affinity:       None
No events.
```

feiskyer committed:
CRI: Add init containers

Commit:
Clean up SchedulerPredicates e2e tests

Reduce repetition & don't use hardcoded pause image tags.

juanvallejo committed:
make portallocator.ErrNotInRange a type

juanvallejo committed:
update known flags

sjenning committed:
kubelet: eviction: allow minreclaim as percentage

Random-Liu committed:
Add seccomp and apparmor support.

0xmichalis committed:
Add a test case and consolidate deployment constructors

Commit:
Enabel custom infracontainer image

Modify api protoc for infra

bprashanth committed:
Enable alpha features in local-up-cluster

vishh committed:
nodefs becomes imagefs on GCI since kubelet cannot identify rootfs correctly

Signed-off-by: Vishnu kannan <vishnuk@google.com>

pmorie committed:
Add node event for container/image GC failure

freehan committed:
try to use ifdown/ifup if available

Random-Liu committed:
Add unit test for get security option functions.

Commit:
EnsureLoadBalancer update instead of recreate existing LBs

guangxuli committed:
if test succ remove the tmp file

Commit:
Make genericapiserver handler chain customizable

ixdy committed:
Create a get-kube-binaries script to download client/server tarballs

Also fix some errors in get-kube.sh, and remove client architectures we
don't officially support.

mikedanese committed:
kubeadm: refactor config

1) break object into substructures
2) seperate a config object for master and node

Commit:
decouple workqueue metrics from prometheus

Commit:
Turn RequestInfoResolver into a RequestInfoFactory

a-robinson committed:
Don't tolerate unready endpoints in cockroachdb example

That was only included in the initial config due to cargo-culting, and
has the potential to break node startup if it resolves its own address
to gossip to.

tallclair committed:
Add the original summary test back

tallclair committed:
Fix time matcher in summary test

derekwaynecarr committed:
Add cgroup driver and cgroups-per-qos flags to kubelet

derekwaynecarr committed:
Add generated code for kubelet flags

nikhiljindal committed:
Enable kubectl describe rs to work when apiserver does not support pods

freehan committed:
move pod networking tests common

jingxu97 committed:
Fix issue in updating device path when volume is attached multiple times

When volume is attached, it is possible that the actual state
already has this volume object (e.g., the volume is attached to multiple
nodes, or volume was detached and attached again). We need to update the
device path in such situation, otherwise, the device path would be stale
information and cause kubelet mount to the wrong device.

This PR partially fixes issue #29324

Commit:
Cosmetical fixes for sysctl psp

Commit:
Complete sysctl PSP tests

Commit:
Add psp sysctl provider tests

Commit:
Return error for empty sysctl names

feiskyer committed:
Kubelet: fix port forward for dockershim

Signed-off-by: Pengfei Ni <feiskyer@gmail.com>

wojtek-t committed:
Remove unnecessary podLister from SelectorSpreading

wojtek-t committed:
Migrate EqualPriority to MapReduce-like framework.

juanvallejo committed:
remove unused GetFlagInt32 func

deads2k committed:
stop plumbing options to start

DirectXMan12 committed:
Add method to inspect Docker images by ID

Previously, the `InspectImage` method of the Docker interface expected a
"pullable" image ref (name, tag, or manifest digest).  If you tried to
inspect an image by its ID (config digest), the inspect would fail to
validate the image against the input identifier.  This commit changes
the original method to be named `InspectImageByRef`, and introduces a
new method called `InspectImageByID` which validates that the input
identifier was an image ID.

deads2k committed:
fix apiserver startup check

DirectXMan12 committed:
Kubelet: Use RepoDigest for ImageID when available

Previously, we used the docker config digest (also called "image ID"
by Docker) for the value of the `ImageID` field in the container status.
This was not particularly useful, since the config manifest is not
what's used to identify the image in a registry, which uses the manifest
digest instead.  Docker 1.12+ always populates the RepoDigests field
with the manifest digests, and Docker 1.10 and 1.11 populate it when
images are pulled by digest.

This commit changes `ImageID` to point to the the manifest digest when
available, using the prefix `docker-pullable://` (instead of
`docker://`)

Random-Liu committed:
Fix kubelet perf data to make work again for perfdash.

deads2k committed:
fix pod eviction storage

madhusudancs committed:
Deregister clusters during federated namespace e2e tear down.

This is causing other tests to leak resources.

Random-Liu committed:
Implement temporary ImageStats in kuberuntime_manager, and
fix a bug in dockershim which causes summary api not working
properly.

Commit:
Update generated openapi

luxas committed:
Set FailureTreshold to 8 for kubeadm components in order to not restart unless really necessary

Random-Liu committed:
Fix bug in dockershim to set sandbox id properly.

sjenning committed:
wait for store update before modify

maxwell92 committed:
add checkPermissions() to ensure kubeadm running in root

maxwell92 commented:
@k8s-bot ok to test

pigmej committed:
Typos and englishify pkg/volume

soltysh committed:
Add asgroups to audit log

pmorie committed:
Proposal: SELinux enhancements

deads2k committed:
fix loopback authorizer

deads2k committed:
fix integration tests for loopback client

0xmichalis committed:
kubectl: return exit status appropriately after running watch.Until

deads2k committed:
specify flags to test-integration

dims committed:
Fix URL in README

Fixes #33888

ixdy committed:
Fix condition for deleting master etcd firewall rule

dashpole committed:
make it push

dashpole committed:
make it push v2

Commit:
CRI: Remove the mount name and port name.

Per discussion on https://github.com/kubernetes/kubernetes/issues/33873.

Currently the mount name is not being used and also involves some
incorrect usage (sometimes it's referencing a mount name, sometimes
it's referecing a volume name), so we decide to remove it from CRI.

The port name is also not used, so remove it as well.

bowei committed:
Add a retry loop to KubeletManagerEtcHosts flake

Note: this still makes the test fail if a retry occurs, but
will give us more information regarding whether or not the
test flake could be occuring due to delay in mounting of
/etc/hosts.

Commit:
discovery client retry serverPreferredResources() in case groups changed in the middle

MrHohn committed:
Fix #33784, IN_CREATE event does not guarantee file content written

jingxu97 committed:
Add a check for file size if the reading content returns empty

In order to debug the flaky tests for writing/reading files via
contains, this PR adds a check for file size if reading returns empty
content.

Commit:
Heal the namespaceless ingresses in federation e2e.

rata committed:
Remove duplicated code in secret e2e tests

This patch just removes duplicated code in secret e2e tests.

freehan committed:
add delete-namespace-on-failure flag

zmerlynn committed:
e2e.go / kops: Add ssh-key option to override ssh key

By default, Jenkins stuffs the ssh key in
/workspace/.aws/kube_aws_rsa. Allow this to be overridden easily on
the command line.

janetkuo committed:
Add status check in PetSet e2e tests

xiang90 committed:
scheduler: cache.delete deletes the pod from node specified in the cached state

jbeda committed:
Remove unused bash variables

wojtek-t committed:
Set deserialization cache size based on target memory usage

luxas committed:
Remove the flannel experimental overlay

luxas committed:
Remove the old flannel network mode in kube-up

luxas committed:
Run hack/update-all.sh

gmarek committed:
Revert "move pod networking tests common"

deads2k committed:
fix watch.until and added listwatch

0xmichalis committed:
hack: run openapi-spec as part of update-all.sh

juanvallejo committed:
Allow empty annotation values

Annotations with empty values can be used, for example, in diagnostics
logging. This patch removes the client-side check for empty values in
an annotation key-value pair.

**Before**
```
$ kubectl annotate pod zookeeper-1 node-selector="" --overwrite
error: invalid annotation format: node-selector=
```

**After**
```
$ kubectl annotate pod zookeeper-1 node-selector="" --overwrite
pod "zookeper-1" annotated
```

```
$ kubectl get po/zookeeper-1 --template='{{.metadata.annotations}}'
map[... node-selector: test-label:test]
```

mwielgus committed:
Add deployment to federation clientset

deads2k committed:
update deployment lister

tallclair committed:
Tweak summary test memory expectations

deads2k committed:
update replicaset lister

MrHohn committed:
examine output format before return, fail the test instead of panic

Commit:
CRI: Update the hack/verify-flags/exceptions.txt

jessfraz committed:
hack: Update cherry-pick script to include original PR subject

Signed-off-by: Jess Frazelle <acidburn@google.com>

zmerlynn committed:
e2e.go / kops: Add --kops-kubernetes-version

This adds a flag to override the version we're deploying with kops for
e2es. (The kops experience is disconnected from the actual source tree
we're running tests from, which is similar to GKE.)

freehan committed:
fix bug in DialFromNode where it fail the test when encountered a command execution failure

freehan committed:
Revert "Revert "move pod networking tests common""

bprashanth committed:
Check for empty string post trimming

Commit:
generated

deads2k committed:
remove testapi.Default.GroupVersion

bprashanth committed:
Add netexec 1.7 to whitelists

Random-Liu committed:
Add presubmit CRI validation test.

Commit:
Add local option to annotate

Commit:
Add local option to the label command

This is to be used with the new "set selector" command.

wojtek-t committed:
Run integration tests with 3.0.10 etcd underneath

wojtek-t committed:
Start testing etcd2->etcd3 migration in test-update-storage-objects

wojtek-t committed:
Make gets for previous value in watch serializable

wojtek-t committed:
Fix call to Index call in listers

0xmichalis committed:
kubectl: add watch flag in rollout status

gmarek committed:
Allow setting short value in integration tests

MrHohn committed:
Fix #31085, include output checking in retry loop

dmmcquay commented:
@maxwell92 This issue is related to #33908 and I was working on it, didn't realize it had been worked on.  I took a slightly different approach since, from what I can tell, returning the error early like that would stop a normal user from even running something like `kubeadm version`, which I don't think we would want that behavior. 

What do you think?

pires commented:
@dmmcquay you should open a PR, since you already have the code, and link it here for comparison.

mwielgus committed:
Daemonset in federated clientset

Random-Liu committed:
Add serial and benchmark test suite.

nikhiljindal committed:
Delete federation namespace after the test completes

liggitt committed:
Fix wait.JitterUntil

errordeveloper commented on cmd/kubeadm/app/kubeadm.go:
```diff
@@ -69,6 +69,17 @@ func Run() error {
+
+func checkPermissions() error {
+	if uid := os.Getuid(); uid != 0 {
+		return fmt.Errorf("Kubeadm needs to run as uid `0`. It is being run as %d", uid)
```
There are a few issues with this error message, first of all it's not exactly user-friendly to talk about UIDs. We should suggest either "run as root" or "use sudo kubeadm".

Secondly, please never capitalise the `kubeadm`, and try to adhere to the rest of error message, we have `<$package/$basename> $message` convention right now.

I'd make it something like:
```
fmt.Errorf("<kubeadm> unable to run as non-root user")
```
The rest of errors come from packages under `app`, so prefix here defers from the convention, but let's discuss. I could be `<main>`, I suppose... We can discuss and improve the convention, of course, but for now let's not break the UX we have.

maxwell92 replied:
Thank you for your suggestions. This is my first time to get involved in such great project. I'll fix my problem : )

maxwell92 replied:
And in my opinion, it could be `<app/kubeadm>`， for the function checkPermissions() is called in function Run() under package `app`.  This structure I learned from `cmd/kubelet/app/options/server.go`

errordeveloper reviewed (changes requested):
Thanks, but please adjust the format of the error message, it's very important 👍 :)

errordeveloper commented:
@dmmcquay @pires oh, I didn't notice your comments (we had another busy day at ContainerCon),  should we close this one?

Commit:
Turn down hootloop logs in priorities.  Hit once cluster is near capacity

mbohlool committed:
Fix a typo in util.sh that resulted in wrong openapi spec generation

mbohlool committed:
Generated openapi spec

Commit:
Update exceptions list for underscores introduced in kubeup for vSphere.

adityakali committed:
Update GCI_VERSION to gci-dev-55-8866-0-0

Changelog:
* Built-in kubernetes updated to v1.4.0
* Enabled VXLAN and IP_SET config options in kernel to support some networking tools
* OpenSSL CVE fixes

Commit:
Add missing argument to log message in federated ingress controller.

wojtek-t committed:
Extend logging in cacher to understand its bottleneck

jbeda committed:
Fix broken build/push-federation-images.sh

This was broken by #30787. A stray bash `source` caused an undefined variable reference.

Apparently the federation images have a parallel nad different "release" path
that isn't tested by the pre-checkin tests.

wojtek-t committed:
Minor clearnup in etcd3 code

gmarek committed:
Small update to scheduler benchmark

dmmcquay commented:
@errordeveloper @pires I think so.  It seems like kubeadm currently uses error handling code from kubectl (k8s.io/kubernetes/pkg/kubectl/cmd/util/helpers.go) and I suggest we create a way to handle kubeadm errors that would be similar to that but specific to kubeadm needs. This involves creating error types throughout kubeadm that can be checked and handled correctly (correct logging, exit code, etc.) instead of jumping out to kubectl to handle them. Seems like we could be more specific with what types of errors come out of kubeadm that way.

I also acknowledge that I have very little experience with the k8s code base and maybe that would be a bad idea.  What do others think?

I think this would help not only with this issue but #33908 and #33928, and probably others.

liggitt committed:
Simplify kubectl get --raw

jingxu97 committed:
Fix nil pointer issue when getting metrics from volume mounter

Currently it is possible that the mounter object stored in Mounted
Volume data structure in the actual state of kubelet volume manager is
nil if this information is recovered from state sync process. This will
cause nil pointer issue when calculating stats in volume_stat_calculator.
A quick fix is to not return the volume if its mounter is nil. A more
complete fix is to also recover mounter object when reconstructing the
volume data structure which will be addressed in PR #33616

mbohlool committed:
Remove unnecessary authorization headers after authorization is successful

pires commented:
@dmmcquay I agree with your assessment. The issues you linked, for instance, clearly show us that we can't rely on `kubectl` error handling, since `kubectl` can run without superuser privileges, etc.

jbeda committed:
Use rsync to probe for rsync ready

We were using netcat to try and figure out when the rsync container is ready. Now we instead use rsync itself. I suspect that there was a race condition with some versions of Docker where it would accept connections and then close them during container start.

This fixes #34214 (I think)

mikedanese committed:
add a test for kubectl apply --prune

ixdy committed:
Make sure rsync.sh is executable inside the build image

Also make sure /etc/localtime is readable

Also move BUILD_IMAGE_VERSION to build-image/VERSION to be consistent
with the way the cross-image works.

errordeveloper commented:
@dmmcquay @pires I agree, and we have an issue open for at least the error codes, as you know (https://github.com/kubernetes/kubernetes/issues/33928). I was thinking `k8s.io/kubernetes/pkg/kubectl/cmd/util` was the state of the art, hence I've used it, but may be it isn't quite right for our needs, I don't know.

0xmichalis committed:
test: bump deployment test timeout

Matches the timeout used in WaitForDeploymentStatus

mwielgus committed:
Race free version of FakeWatcher

mwielgus committed:
Use RaceFreeFakeWatcher in federation controllers helper and ingress controller

Commit:
build: fix typo in error message help

"hasn't been set of is set" --> "hasn't been set or is set"

dmmcquay commented:
@errordeveloper @pires I'll put something together with how I think it should be and then you can decide if you think its a good idea.

Commit:
Revert "Add kubelet awareness to taint tolerant match caculator."

jessfraz committed:
Update CHANGELOG.md for v1.4.1-beta.2.

maxwell92 committed:
add checkPermissions()

maxwell92 commented:
@dmmcquay @pires @errordeveloper thank you for your attentions. I'm fresh to this. Seems like I did something wrong in my contribution..

maxwell92 commented:
maybe I should first learn how to git rebase @errordeveloper

errordeveloper commented:
@maxwell92 looks like we have this underway in https://github.com/kubernetes/kubernetes/pull/34341, and I'm afraid that means we will need to close this. // cc @luxas

luxas commented:
Yes, unfortunately I think so.

@maxwell92 Anything other you're willing to contribute?
There are lots of open issues.

And yeah, this really needs a rebase.
